lib/xandra/connection.ex

```diff
@@ -230,17 +237,80 @@ defmodule Xandra.Connection do
     {:ok, batch, state}
   end
 
+  defp send_prepared(%Prepared{compressor: compressor} = prepared, socket, options, state) do
+    transport = state.transport
+
+    frame_options =
+      options
+      |> Keyword.take([:tracing, :custom_payload])
+      |> Keyword.put(:compressor, compressor)
+
+    payload =
+      Frame.new(:prepare, frame_options)
+      |> state.protocol_module.encode_request(prepared)
+      |> Frame.encode(state.protocol_module)
+
+    protocol_format = Xandra.Protocol.frame_protocol_format(state.protocol_module)
+
+    with :ok <- transport.send(socket, payload),
+         {:ok, %Frame{} = frame} <-
+           Utils.recv_frame(transport, socket, protocol_format, state.compressor) do
+      frame = %Frame{frame | atom_keys?: state.atom_keys?}
+
+      case state.protocol_module.decode_response(frame, prepared) do
+        {%Prepared{} = prepared, warnings} ->
+          maybe_execute_telemetry_event_for_warnings(state, prepared, warnings)
+          Prepared.Cache.insert(state.prepared_cache, prepared)
+          {:ok, prepared, state}
+
+        %Xandra.Error{} = reason ->
+          {:error, reason, state}
+      end
+    else
+      {:error, reason} ->
+        {:disconnect, ConnectionError.new("prepare", reason), state}
+    end
+  end
+
   @impl true
   def handle_execute(query, payload, options, %__MODULE__{} = state) do
-    %{socket: socket, compressor: compressor, atom_keys?: atom_keys?} = state
-    assert_valid_compressor(compressor, options[:compressor])
+    assert_valid_compressor(state.compressor, options[:compressor])
+
+    telemetry_metadata = Keyword.fetch!(options, :telemetry_metadata)
+
+    metadata = %{
+      query: query,
+      connection_name: state.connection_name,
+      host: state.address,
+      port: state.port,
+      extra_metadata: telemetry_metadata
+    }
+
+    :telemetry.span(
+      [:xandra, :execute_query],
+      metadata,
+      fn ->
+        case send_query(query, payload, options, state) do
+          {:ok, response, state} ->
+            {{:ok, query, response, state}, metadata}
 
+          {:error, reason, state} ->
```

Comment:
**Nitpick**: let's make sure this has the correct name, and some extra security by matching on it.

```suggestion
          {:error, %Xandra.Error{} = error, state} ->
```